Add bedrock/signals accelerator (#1680)
snowplow/documentation: tutorials/signals-agentic-accelerator/add-memory.md

---
title: "Add persistent customer context with AgentCore Memory"
sidebar_label: "Add agent memory"
position: 4
description: "Integrate AWS Bedrock AgentCore Memory to give your AI agent persistent context across conversations using preference and semantic memory strategies."
keywords: ["AgentCore Memory", "persistent memory", "personalization", "hooks", "Strands Agents", "long-term memory"]
date: "2026-03-27"
---

Without persistent memory, every conversation with an AI agent starts from zero. Users must repeat their preferences, explain their history, and re-establish context each time.

> [AgentCore Memory](https://docs.aws.amazon.com/bedrock-agentcore/latest/devguide/memory.html) solves this by providing managed short-term and long-term memory for agents. In this step you create a memory resource, seed it with sample customer history, and integrate it into your agent using hooks that automatically retrieve and store context.

Reviewer: ```suggestion
[AgentCore Memory](https://docs.aws.amazon.com/bedrock-agentcore/latest/devguide/memory.html) solves this by providing managed short-term and long-term memory for agents. In this step, you'll create a memory resource, seed it with sample customer history, and integrate it into your agent using hooks that automatically retrieve and store context.
```

## Create AgentCore Memory resources

Run the notebook cell that creates a memory resource with two strategies:

| Strategy | Purpose | Namespace |
| :------- | :------ | :-------- |
| `USER_PREFERENCE` | Learns customer preferences and behavioral patterns | `travel/customer/{actorId}/preferences` |
| `SEMANTIC` | Captures factual information and conversation context | `travel/customer/{actorId}/semantic` |

> The `{actorId}` placeholder in the namespace is replaced at runtime with the customer identifier, ensuring each customer's memories are isolated.

Reviewer: ```suggestion
The `{actorId}` placeholder in the namespace will be replaced at runtime with the customer identifier, ensuring each customer's memories are isolated.
```

where's this defined?

Author: In the next step

```
memory_client.create_event(
    memory_id=memory_id,
    actor_id=CUSTOMER_ID,
    session_id="previous_session",
    messages=previous_interactions,
)
```

```python
response = memory_client.create_memory_and_wait(
    name="TravelAgentMemory",
    description="Travel agent memory",
    strategies=strategies,
    event_expiry_days=90,
)
```

> :::warning

Reviewer: ```suggestion
:::note[Creation time]
```

Memory resource creation can take a couple of minutes. Wait for the confirmation message before proceeding.
:::

## Seed customer history

Run the notebook cell that loads sample previous interactions. This simulates a returning customer who has expressed preferences for warm, adventurous, family-friendly destinations:

```python

> previous_interactions = [

Reviewer: missing the CUSTOMER_ID definition line

    ("I'd love to go and visit somewhere warm.", "USER"),

>     ("I can help with that! For warm destinations, I'd recommend Bali, "

Reviewer: the broken strings here look odd. is this the right syntax?

Author: This is just a python preference to keep line length shorter

* `retrieve_customer_context` - fires when a new user message arrives, retrieves relevant memories from both namespaces, and injects them into the message as context
* `save_travel_interaction` - fires after the agent responds, saves the interaction to memory for future retrieval

The `retrieve_memories` call uses semantic search - it finds memories relevant to the current query, not just the most recent ones. This means if a user asks about beach destinations, the agent retrieves memories about their beach preferences even if those were mentioned weeks ago.

## Test personalized responses

Run the notebook cells that rebuild the agent with memory hooks and test it. The agent now has access to both memory (the seeded customer history) and Signals (the behavioral attributes from your test events).

```python
agent_with_memory = Agent(
    model=model,
    hooks=[memory_hooks],
    tools=[get_destination_info, get_experience_info, web_search, get_signals],
    system_prompt=SYSTEM_PROMPT,
)
```

> Test with queries that should trigger memory retrieval:

Reviewer: include example output

```python
response = agent_with_memory("Can you recommend some destinations for me?")
```

Because you seeded `customer_001` with preferences for warm, adventurous, family-friendly destinations, the agent should incorporate this context into its recommendations without the user needing to restate those preferences.

To test Signals integration, ask the agent a question that triggers the `get_signals` tool:

```python
response = agent_with_memory(

>     "Based on my browsing behavior, what experiences would suit me?"

Reviewer: this feels clunky, surely no one will ever say something like this in reality. fine for an example, but maybe say below that in production Signals tool usage will be automatic?

also, provide example output
